src/acc/libsmm_acc/libcusmm/predict_helpers.py

```diff
@@ -0,0 +1,223 @@
+# -*- coding: utf-8 -*-
+####################################################################################################
+# Copyright (C) by the DBCSR developers group - All rights reserved                                #
+# This file is part of the DBCSR library.                                                          #
+#                                                                                                  #
+# For information on the license, see the LICENSE file.                                            #
+# For further information please visit https://dbcsr.cp2k.org                                      #
+# SPDX-License-Identifier: GPL-2.0+                                                                #
+####################################################################################################
+
+import sys
+import os
+import pickle
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+from kernels.cusmm_predict import to_string
+
+
+# ===============================================================================
+# I/O helpers
+def safe_pickle(data, file):
+    """
+    Pickle big files safely by processing them in chunks
+    :param data: data to be pickled
+    :param file: file to pickle it into
+    """
+    max_bytes = 2**31 - 1  # Maximum number of bytes to pickle in one chunk
+    pickle_out = pickle.dumps(data)
+    n_bytes = sys.getsizeof(pickle_out)
+    with open(file, "wb") as f:
+        count = 0
+        for i in range(0, n_bytes, max_bytes):
+            f.write(pickle_out[i:min(n_bytes, i + max_bytes)])
+            count += 1
+
+
+def safe_pickle_load(file_path):
+    max_bytes = 2**31 - 1
+    bytes_in = bytearray(0)
+    input_size = os.path.getsize(file_path)
+    with open(file_path, "rb") as f:
+        for _ in range(0, input_size, max_bytes):
+            bytes_in += f.read(max_bytes)
+    return pickle.loads(bytes_in)
```

Comment:
Can you elaborate on what kind of problem you're trying to solve here and why `pickle.dump`/`pickle.load` does not work?

The reading part looks rather inefficient and could probably benefit from `mmap` if `pickle.load` can't cope with large amount of data, without having to manually build a `bytearray` with the complete data:

```
import mmap
from contextlib import contextmanager

@contextmanager
def mmap_open(name, mode='r'):
    access = mmap.ACCESS_READ if mode == 'r' else mmap.ACCESS_WRITE
    with open(name, mode + 'b') as fhandle:
        fmapped = mmap.mmap(fhandle.fileno(), 0, access=access)
        yield fmapped
        fmapped.close()

with mmap_open(fpath) as fmapped:
    return pickle.loads(fmapped)
```

Reply:
`pickle.dump`/`load` with Python3 on OSX has trouble handling very large data (see https://bugs.python.org/issue24658). These two functions are meant as a safety wrapper around `pickle.dump`/`load` to avoid this problem.

Reply:
I'm not too concerned about the performance of the load since this function is called only in the beginning of the training (`predict_train.py`) (optionally, for evaluating a pre-trained model) and the parameter generation (`predict_genpars`), both long procedures which would not be impacted much by reducing the time of loading they do once. 

I'd prefer sticking with a simpler and more readable package.